[master] Add a minimal SSH test.pkg_opts minion cache.

Rudd-O opened the pull request:
This only attempts to accelerate repeated contacts with the SSH minion. The amount of data formerly cached is not contemplated in this patch, due to the complexity of ensuring that all that data is invalidated.

### What does this PR do?

Accelerates repeated contacts with SSH minions by roughly 1 second, or more if latency is high.

Halves the total amount of SSH connections established with SSH minions.

In particular, orchestration runs become faster.

Most of the change is indentation whitespace.

### What issues does this PR fix or reference?

None.

### Previous Behavior

Every SSH minion run was prefaced with a `test.opts_pkg` call, slowing down minion runs.

### New Behavior

SSH minion runs recently contacted  to obtain the `test.opts_pkg`
data are not contacted for that data anymore.  Instead, this data is read from a local minion
cache file, which is properly locked to protect data integrity.

This is by default turned off.

### Merge requirements satisfied?

**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] No tests have been updated -- please point me in the right direction.

### Commits signed with GPG?

Yes.

Rudd-O committed:
Partially restore SSH minion cache.

This only attempts to accelerate repeated contacts with the SSH minion.
The amount of data formerly cached is not contemplated in this patch,
due to the complexity of ensuring that all that data is invalidated.

Rudd-O commented on salt/client/ssh/__init__.py:
```diff
@@ -1206,95 +1340,121 @@ def run_wfunc(self):
         if "ssh_grains" in self.opts:
             conf_grains = self.opts["ssh_grains"]
-        if not data_cache:
-            refresh = True
```
This is just an indentation change!

twangboy reviewed (changes requested):
This will also need a test

dwoz committed:
Fix pylint errors and RST formatting in SSH minion cache patch

- Add pylint: disable=resource-leakage on fopen call held outside with block
- Add pylint: disable=broad-except on three bare except Exception clauses
- Remove constant `if True:` guard (W0125) and de-indent the enclosed block
- Fix RST: add missing .. conf_master:: ssh_minion_opts_cache directive,
  lengthen underline from 23 to 24 dashes, restore .. conf_master:: thin_extra_mods

dwoz committed:
Add unit tests for cache_context_manager in salt.client.ssh

Tests cover: nocache fallback when path is None or fcntl unavailable;
filecache persist/retrieve round-trip; invalidate wipes content;
older_than threshold logic; and deduplication (no write when retrieved
payload is unchanged).

twangboy reviewed (changes requested):
Since this introduces a brand new master configuration option (ssh_minion_opts_cache), it would be great to make it easily discoverable for users by adding it to the default configuration template.

Could we update salt/config/__init__.py to include the default value in the master configuration block? Something like:

```python
# In salt/config/__init__.py under the master configuration defaults:
"ssh_minion_opts_cache": 0,
```

That way, it's explicitly documented within the codebase's default options alongside the documentation update you already added to master.rst.
